module/zfs/zfs_quota.c

```diff
@@ -350,15 +351,29 @@ zfs_set_userquota(zfsvfs_t *zfsvfs, zfs_userquota_prop_t type,
 		VERIFY(0 == zap_add(zfsvfs->z_os, MASTER_NODE_OBJ,
 		    zfs_userquota_prop_prefixes[type], 8, 1, objp, tx));
 	}
-	mutex_exit(&zfsvfs->z_lock);
 
 	if (quota == 0) {
 		err = zap_remove(zfsvfs->z_os, *objp, buf, tx);
 		if (err == ENOENT)
 			err = 0;
+		/*
+		 * If the quota contains no more entries after the entry
+		 * was removed, destroy the quota zap and remove the
+		 * reference from zfsvfs. This will save us unnecessary
+		 * zap_lookups for the quota during writes.
+		 */
+		VERIFY(0 == zap_count(zfsvfs->z_os, *objp, &zap_nentries));
+		if (zap_nentries == 0) {
+			prev_objp = *objp;
```

Comment:
It looks like `prev_objp` isn't really needed and `*objp` could be passed directly to `zap_destroy()`.  Optionally you could then set it to zero afterwards, although since it's unused beyond that point it's strictly not needed.

Reply:
Ah, yeah. Was worried about the safety of destroying the obj before its reference it removed in zfsvfs but we're inside a zfsvfs lock anyway.